index.js.flow

```diff
@@ -23,69 +20,60 @@ declare export function isArrayLike(obj: any): boolean
-  collection: TCollection,
-  callbackFn: (value: mixed, index: number, collection: TCollection) => any,
-  thisArg?: any
-): void
+declare export var getIterator:
+  & (<TValue>(iterable: Iterable<TValue>) => Iterator<TValue>)
+  & ((iterable: mixed) => void | Iterator<mixed>)
+
+declare export var getIteratorMethod:
+  & (<TValue>(iterable: Iterable<TValue>) => (() => Iterator<TValue>))
+  & ((iterable: mixed) => (void | (() => Iterator<mixed>)))
+
+declare export var createIterator:
+  & (<TValue>(collection: Iterable<TValue>) => Iterator<TValue>)
+  & ((collection: {length: number}) => Iterator<mixed>)
+  & ((collection: mixed) => (void | Iterator<mixed>))
+
+declare export var forEach:
+  & (<TValue, TCollection: Iterable<TValue>>(
+      collection: TCollection,
+      callbackFn: (value: TValue, index: number, collection: TCollection) => any,
+      thisArg?: any
+    ) => void)
+  & (<TCollection: {length: number}>(
+      collection: TCollection,
+      callbackFn: (value: mixed, index: number, collection: TCollection) => any,
+      thisArg?: any
+    ) => void)
 
 declare export var $$asyncIterator: '@@asyncIterator'
 
 declare export function isAsyncIterable(obj: any): boolean
 
-declare export function getAsyncIterator<TValue>(
-  asyncIterable: AsyncIterable<TValue>
-): AsyncIterator<TValue>
-
-declare export function getAsyncIteratorMethod<TValue>(
-  asyncIterable: AsyncIterable<TValue>
-): () => AsyncIterator<TValue>
-
-declare export function createAsyncIterator<TValue>(
-  collection: AsyncIterable<TValue> | Iterable<Promise<TValue> | TValue>
-): AsyncIterator<TValue>
-declare export function createAsyncIterator(collection: {
-  length: number
-}): AsyncIterator<mixed>
-
-declare export function forAwaitEach<
-  TValue,
-  TCollection: AsyncIterable<TValue>
->(
-  collection: TCollection,
-  callbackFn: (value: TValue, index: number, collection: TCollection) => any,
-  thisArg?: any
-): Promise<void>
-declare export function forAwaitEach<
-  TValue,
-  TCollection: Iterable<Promise<TValue> | TValue>
->(
-  collection: TCollection,
-  callbackFn: (value: TValue, index: number, collection: TCollection) => any,
-  thisArg?: any
-): Promise<void>
-declare export function forAwaitEach<TCollection: { length: number }>(
-  collection: TCollection,
-  callbackFn: (value: mixed, index: number, collection: TCollection) => any,
-  thisArg?: any
-): Promise<void>
+declare export var getAsyncIterator:
+  & (<TValue>(asyncIterable: AsyncIterable<TValue>) => AsyncIterator<TValue>)
+  & ((asyncIterable: mixed) => (void | AsyncIterator<mixed>))
+
+declare export var getAsyncIteratorMethod:
+  & (<TValue>(asyncIterable: AsyncIterable<TValue>) => (() => AsyncIterator<TValue>))
+  & ((asyncIterable: mixed) => (void | (() => AsyncIterator<mixed>)))
+
+declare export var createAsyncIterator:
+  & (<TValue>(
+      collection:
+        | Iterable<Promise<TValue> | TValue>
+        | AsyncIterable<TValue>
```

Comment:
@nmn - I was able to get this to work. The issue wasn't that `string` isn't being detected as `{length: number}` - really we wanted to detect it as `Iterable<string>` - the same issue with Array being being correctly detected as `Iterable<string>`. I was able to fix it by flipping the order of this union type - now it checks for `Iterable` before it checks for `AsyncIterable`. I'm not 100% sure why this works, but it does.

What I expected was the `AsyncIterable` line to not match and have flow then try to match the `Iterable` line and have it match for Arrays and strings.